### ISSUE TITLE
When inside enclave the digest generated for some files are different compared to non-enclave environment

### ISSUE THREAD

**Issue by democ98** [gramineproject/gramine#1458]
### Description of the problem

I have a method written in rust that runs in the enclave. Its main function is:
1. Generate an 8MB file and store it in the '/sgx/' path.
2. Call an algorithm and pass in the path of the file. The algorithm will generate a file digests according to the file content (note: whenever the file content is the same, the digests content will also be the same. When running this algorithm, the CPU basically is full load)

My approach is:
* Inside enclave, when I called the modification method 256 times and generated 256 digests. I verified these digests using the verification method corresponding to the algorithm, and the result was failure.
* Use the same algorithm to generate 256 digests for these 256 8MB files that have been generated in a non-enclave environment, and verifiying them is successful with the same verification method.

My findings:
After comparison, it is found that there are more than 20 inconsistent values in the digests lists generated twice, which is the reason for the verification failure.

My guess:
Some methods read the file content, and the wrong file content has been read.

### Steps to reproduce

Here's the outermost loop:
```rust
let mut times = 0;
while times < 256 {
	try_create_tag();
	times+=1;
}
```

The code to call the calculation after generating the random file:
```rust
//Generate random file
let mut rng = rand::thread_rng();

/*pseudocode*/
let mut file_path = "/sgx/".to_string();
let file_name = /*some random function*/;
file_path = file_path.clone() + &format!("{:?}", file_name);
let mut test_file_handle = match std::fs::File::create(file_path.clone()){
    .../*error handling*/...
};

let block_size = 2097152; /* 2MB */
let blocks:usize = 4;
for _ in 0..blocks {
	let mut rng_data = vec![0u8; block_size];
	rng.fill_bytes(&mut rng_data);
	hasher.input(&rng_data);
	match test_file_handle.write_all(&rng_data) {
		.../*error handling*/...
	};
	match test_file_handle.flush() {
		.../*error handling*/...
	};
}
match test_file_handle.sync_all(){
	.../*error handling*/...
};

//Algorithms for Computing Digests
let pool = ThreadPool::new(/*max thread,my is 12*/)
let digests= match sig_gen(&file_path, pool.clone(),1024,/*Other irrelevant fixed parameters*/)
{
    .../*error handling*/...
};

/*save digests in somewhere in order*/
```

Specific algorithm content:

```rust
//Running the algorithm will use some exponential operations, and the CPU will always run at 100%
pub fn sig_gen(
		file_path: &String,
		pool: ThreadPool,
    	n_blocks: usize,/*Compute the digest for the 8MB file in how many chunks*/
    	...
	) -> Result<Tag, Error> {
       	///I only show where a file is read and then a digest is calculated based on the file content
        //open file
		let mut f = fs::File::open(file_path).map_err(....)?;
        //detect file size
        let file_size:usize = match f.seek(SeekFrom::End(0)) {
            ...
        };
		if file_size % n_blocks != 0 {
			return Err(Error {...})
		};
         let each_chunks_size = file_size / n_blocks;
		let mut offset = 0_u64;
        
        
        let (tx, rx) = channel();
        for i in 0..n_blocks {
            Read the file chunk by chunk
            let mut chunks = vec![0u8; each_chunks_size];
            let _ = f.seek(SeekFrom::Start(offset));
            match f.read_exact(&mut chunks) {
                .../*error handling*/...
            };
            
            let tx = tx.clone();
            //Compute digests for each chunks using a thread pool
            pool.execute(move || {
                tx.send((generate_sigma(chunks), i)).unwrap();
            });
            offset += each_chunks_size;
        }
        
        //Each file chunk digest is put in order
        let iter = rx.iter().take(n_blocks as usize);
        for k in iter {
            digests.phi[k.1 as usize] = k.0.clone();
        }
        
        /*Finally, the value of phi will be aggregated, similar to the Merkle hash tree create hashroot*/
        Ok(digests)
}
```

Also this is my manifest file content:
```shell
[fs]
mounts = [
  { path = "/lib", uri = "file:{{ gramine.runtimedir() }}" },
  { path = "{{ arch_libdir }}", uri = "file:{{ arch_libdir }}" },
  { path = "/usr{{ arch_libdir }}", uri = "file:/usr{{ arch_libdir }}" },
  { path = "/sgx", uri = "file:/sgx" }
]
[sgx]
allowed_files = [
   ...
  "file:/sgx",
]
...

### Expected results

Two digest calculations for the same file list content should be exactly the same

### Actual results

Basically the same, but there are a few differences.

### Gramine commit hash

gramineproject/gramine digest:7e5b8b6111a6

**Comment by dimakuv** [comment]
1. This app is too specific and fragmented. Could you create a small reproducer that we could easily play with?
2. What happens if you specify a ThreadPool with a single thread, instead of 12?
